src/UserSpaceInstrumentation/Trampoline.cpp

```diff
@@ -0,0 +1,155 @@
+                                                     const AddressRange& range) {
+  for (size_t i = 0; i < ranges_sorted.size(); i++) {
+    if (DoAddressRangesOverlap(ranges_sorted[i], range)) {
+      return i;
+    }
+  }
+  return std::nullopt;
+}
+
+std::optional<size_t> HighestIntersectingAddressRange(
+    const std::vector<AddressRange>& ranges_sorted, const AddressRange& range) {
+  for (int i = ranges_sorted.size() - 1; i >= 0; i--) {
+    if (DoAddressRangesOverlap(ranges_sorted[i], range)) {
+      return i;
+    }
+  }
+  return std::nullopt;
+}
+
+ErrorMessageOr<std::vector<AddressRange>> GetTakenAddressRanges(pid_t pid) {
+  std::vector<AddressRange> result;
+  OUTCOME_TRY(mmap_min_addr, ReadFileToString("/proc/sys/vm/mmap_min_addr"));
+  uint64_t mmap_min_addr_as_uint64 = 0;
+  if (!absl::SimpleAtoi(mmap_min_addr, &mmap_min_addr_as_uint64)) {
+    return ErrorMessage("Failed to read/proc/sys/vm/mmap_min_addr");
+  }
+  result.emplace_back(0, mmap_min_addr_as_uint64);
+
+  OUTCOME_TRY(maps, ReadFileToString(absl::StrFormat("/proc/%d/maps", pid)));
+  std::vector<std::string> lines = absl::StrSplit(maps, '\n', absl::SkipEmpty());
+  for (const auto& line : lines) {
+    std::vector<std::string> tokens = absl::StrSplit(line, ' ', absl::SkipEmpty());
+    if (tokens.size() < 1) {
+      continue;
+    }
+    std::vector<std::string> addresses = absl::StrSplit(tokens[0], '-');
+    if (addresses.size() != 2) {
+      continue;
+    }
+    uint64_t address_begin = 0;
+    uint64_t address_end = 0;
+    if (!safe_strtou64_base(addresses[0], &address_begin, 16) ||
+        !safe_strtou64_base(addresses[1], &address_end, 16)) {
+      continue;
+    }
+    CHECK(address_begin < address_end);
+    // Join with previous segment ...
+    if (result.back().end == address_begin) {
+      result.back().end = address_end;
+      continue;
+    }
+    // ... or append as new segment.
+    result.emplace_back(address_begin, address_end);
+  }
+  return result;
+}
+
+ErrorMessageOr<AddressRange> FindAddressRangeForTrampoline(
+    const std::vector<AddressRange>& taken_ranges, const AddressRange& code_range, uint64_t size) {
+  constexpr uint64_t kMax32BitOffset = 0x7fffffff;
+  const uint64_t page_size = sysconf(_SC_PAGE_SIZE);
+
+  // Try to fit an interval of length `size` below `code_range`.
+  auto optional_range_index = LowestIntersectingAddressRange(taken_ranges, code_range);
+  if (!optional_range_index) {
+    return ErrorMessage(absl::StrFormat("code_range %#x-%#x is not in taken_ranges.",
+                                        code_range.start, code_range.end));
+  }
+  while (optional_range_index.value() > 0) {
+    // Place directly to the left of the take interval we are in...
+    uint64_t address_trampoline = taken_ranges[optional_range_index.value()].start - size;
+    // ... but round down to the previous page boundary.
+    address_trampoline = (address_trampoline / page_size) * page_size;
```

Comment:
completely optional: page_size is a power of 2 so you can replace it with something like address_trampoline & ~(page_size - 1)

Reply:
Hmm, true, but I'll keep as is since it is mildly more readable imo.